### PR TITLE
Releases/v0.20.0

### PR DESCRIPTION
**Pull request by ylvaselling**
This is the new version of the documentation that will be released for version 20.

To test:
* Code review `generatedocs.py`. Feedback is welcome since I don't code in python that much.
* Build the documentation and view it (see below for instructions). Feedback on anything is welcome, for example
    - Design
    - Structure
    - How well the current structure works for our documents 
    - Try it in dark and light mode (toggle this in the web inspector)

(A lot of the choices have been a joint effort, so some comments might not be able to act upon as we would have to discuss any changes)

Note that using the documentation locally is not going to show you the details in the search result list.

Some questions:
* Do we want to merge the RenderableOverview page before we have the Jenkins image tests?
* Do we want to merge the json files generated from the DocsWriter?
* What to do with the `About` section? It was decided to remove this section, but we haven't resolved where to put the content so I just added the section last for now.
* Do we want to remove empty pages? There are some placeholders

Building steps:
  1. In a shell, move to this directory.
  2. Run `python -m venv .venv` to create a python virtual environment.
  3. Run `.\.venv\Scripts\activate` to activate the python virtual environment.
  4. `pip install -r requirements.txt` (this is technically only necessary for the first time or when the dependencies change, but it is a good habit to run this every time after a checkout).
  5. Run `make html`.

### CODE REVIEW

File: generatedocs.py [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+    # This will ensure the simplest asset is displayed
+    assetFiles.sort(key=getFileLength)
+
+    # Find example for the asset component
+    for assetFile in assetFiles:   
+        # Search for Type = "<name>"
+        regex = r'Type = \"' + name + r'\"'
+        example = getLinesAndContentFromFile(assetFile, regex)
+        # As soon as we have a match, we return it
+        # It will be the shortest assets since they are sorted on length
```
Comment (WeirdRubberDuck) [inline generatedocs.py:126]: Coments here seems a bit wordy for such a short function. 

```suggestion
    # Find first example matching the asset component (files are sorted by length)
    for assetFile in assetFiles:   
        # Search for Type = "<name>"
        regex = r'Type = \"' + name + r'\"'
        example = getLinesAndContentFromFile(assetFile, regex)
```

File: generatedocs.py [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+    optionalIndices = []
+    for i, member in enumerate(members):
+        if member["optional"]:
+            optionalIndices.append(i)
+        else:
+            nonOptionalIndices.append(i)
+    indices = nonOptionalIndices + optionalIndices
+    return [members[i] for i in indices]
+
+def findAssetScreenshot(name):
```
Comment (WeirdRubberDuck) [inline generatedocs.py:177]: Since all the other funciton have a describing comment, maybe this one should too?

File: generatedocs.py [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+                if image:
+                    images[assetComponent["name"]] = image
+                renderables.append(assetComponent)
+
+    # Create overview file
+    renderableOverviewTemplate = environment.get_template("renderableOverviewTemplate.txt")
+    outputOverview = renderableOverviewTemplate.render(folderNameAssets=folderNameAssets,
+                                                       renderables=renderables, 
+                                                       images=images
+                                                       )
```
Comment (WeirdRubberDuck) [inline generatedocs.py:380]:
```suggestion
    outputOverview = renderableOverviewTemplate.render(
        folderNameAssets=folderNameAssets,
        renderables=renderables, 
        images=images
    )
```

File: generatedocs.py [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+    outputIndex = indexScriptingTemplate.render(libraries=scriptingApi)
+    with open(os.path.join(scriptingOutputPath, "index.md"), 'w') as f:
+        f.write(outputIndex)
+
+################################################################################
+#                         CREATE RENDERABLE OVERVIEW                           #
+################################################################################
+
+def generateRenderableOverview(environment, outputFolder, folderNameAssets, jsonLocation):
+    # Open JSON file
```
Comment (WeirdRubberDuck) [inline generatedocs.py:354]: The other functions start by creating a target folder. Why does this one not need to?

File: index.md [hunk 5]
```diff
@@ -39,117 +39,119 @@ The overall structure of the documentation is as follows:
...
-support/kitchensink
-support/field-list-example
+about/academics
+about/releases/index
+about/links
+
 :::
+
+
+
```
Comment (WeirdRubberDuck) [inline index.md:157]:
```suggestion
```

File: using-openspace/scripting/index.md [hunk 6]
```diff
@@ -0,0 +1,10 @@
+# Scripting
+
+
```
Comment (WeirdRubberDuck) [inline using-openspace/scripting/index.md:3]:
```suggestion
```

File: using-openspace/scripting/index.md [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+```{toctree}
+:hidden:
+
+console/index
+
```
Comment (WeirdRubberDuck) [inline using-openspace/scripting/index.md:8]:
```suggestion
```

File: using-openspace/scripting/index.md [hunk 8]
```diff
@@ -0,0 +1,10 @@
+# Scripting
+
```
Comment (WeirdRubberDuck) [inline using-openspace/scripting/index.md:2]: This page should have some top level information, but maybe that's for a later stage?

**Review by WeirdRubberDuck — CHANGES_REQUESTED** [review]
👍 Some comments aside from the one in code:

1. Did not manage to enable dark mode, even if I put my browser in dark mode (Chrome). 
1. Building the page creates a lot of .md files that now appear as changed files. These should be added to .gitignore
1. Is this considered the finished-finished version for 0.20.0?  Should it relaly be merged into the `master` branch then?
1. Also, regarding the placeholder pages, if this is the finished-finished version then we should try to fill a few of them with content.  Like the `Community` page

1. The order of the navigation list changes when I navigate through the page. EI have seen "About" move to the top of the list if I for example navigate to "Getting Started -> Hardware requirements"
1. A not for future work: We still need to go through all the pages and make sure we use consistent title casing everywhere. E.g. note how "Handling Errors" and "Hardware requirements" under getting started has different title casing

1. Design wise, it would be nice to have some more indentaiton fo rhte subitems in the navigation menu to the left. See exmaple below .It's difficult to see that Development is a category under "Contributing". (This image also shows some structural issues that we should fix...)

    image

    Version indentation on Godot page, which to me is clearer and easier to distinguish visually, for comparison
    image



1. On my screen resolution, when viewed in 100%, I find the webpage a bit difficult to read due to the text taking up all the width. How about reducing the width of the page content, like for example Godot does?  This makes the content much easier to overview and read. But we could probably make it a little bit wider than them. https://docs.godotengine.org/en/stable/about/faq.html 

    - See pic below for how it looks on my display. Also note the double "Navigation" in the Getting started menu (these are two different pages, but looks a bit weird IMO)
    
    
    image

**Comment by alexanderbock** [comment]
For 1, I think having a button somewhere to toggle between light, dark, and auto mode similar to the current version would be best I believe

3. We'd want to merge this branch into `master` but then either keep this branch around or recreate it once we push the version out.  Functionally the same as leaving it as a separate branch, but getting some comments on this PR is worth the little detour I think.  So no, it's not the finished-finished version yet, but a big step towards it

8. That was my desire to use 100% of the screen width. If that is too wide, then the browser should be thinner instead ;)  I don't really like how "modern" webpage waste a lot of the screen space.

File: .gitignore [hunk 9]
```diff
@@ -1 +1,2 @@
 ﻿build
+.venv
```
Comment (alexanderbock) [inline .gitignore:2]:
```suggestion
.venv

```

File: templates/assetComponentTemplate.txt [hunk 10]
```diff
@@ -0,0 +1,89 @@
+{% macro memberTable(inputMembers, parentName="", linkDescription=true) -%}
```
Comment (alexanderbock) [inline templates/assetComponentTemplate.txt:1]: For this and the other templates, Jinja recommends `.html.jinja` as an extension for template files:
https://jinja.palletsprojects.com/en/3.0.x/templates/#template-file-extension

It think it would be more understandable to use their extension instead of `.txt`

File: conf.py [hunk 11]
```diff
@@ -1,5 +1,11 @@
 import os
 
+# opening file_1.py and reading it with read() and executing if with exec()
+with open("generatedocs.py", encoding="utf8") as file:
+    try:
+      exec(file.read())
+    except Exception as e: 
+      print(e)
```
Comment (alexanderbock) [inline conf.py:8]:
```suggestion
  try:
    exec(file.read())
  except Exception as e: 
    print(e)
```

File: generatedocs.py [hunk 12]
```diff
@@ -0,0 +1,413 @@
+from git import Repo # git for downloading assets
```
Comment (alexanderbock) [inline generatedocs.py:1]: Getting:
`No module named 'git'`
when running `make html`.  The requirements.txt mentions GitPython instead of Git, so maybe it is missing the actual dependency?

File: generatedocs.py [hunk 13]
```diff
@@ -0,0 +1,413 @@
...
+    # Download assets files from OpenSpace repository
+    assetsFolder = cloneAssetsFolderGit(os.path.join(outputFolder, assetsExamplesFolderName))
+
+    # Create target folder
+    assetsOutputPath = os.path.join(outputFolder, folderNameAssets)
+    if not os.path.exists(assetsOutputPath):
+        os.mkdir(assetsOutputPath)
+
+    # Open JSON file
+    f = open(os.path.join(jsonLocation, 'assetComponents.json'))
```
Comment (alexanderbock) [inline generatedocs.py:240]:
```suggestion
    f = open(os.path.join(jsonLocation, "assetComponents.json"))
```

File: generatedocs.py [hunk 14]
```diff
@@ -0,0 +1,413 @@
...
+def parseDoxygenComments(library):
+    for function in library["functions"]:
+        [helpText, p, returnDescription] = function["help"].partition('\\\\return')
+        
+        # Parse code blocks
+        helpText = helpText.replace('\\\\code', '\n:::{code-block} lua\n')
+        helpText = helpText.replace('\\\\endcode', '\n:::')
+
+        # Split help text into parameters and return type
+        helpText = helpText.split('\\\\param ')
```
Comment (alexanderbock) [inline generatedocs.py:200]:
```suggestion
        helpText = helpText.split("\\\\param ")
```

File: generatedocs.py [hunk 15]
```diff
@@ -0,0 +1,413 @@
...
+
+# This function modifies the scripting library so that the doxygen 
+# comments are added to the arguments
+# Supported doxygen parameters: \param \return \code
+def parseDoxygenComments(library):
+    for function in library["functions"]:
+        [helpText, p, returnDescription] = function["help"].partition('\\\\return')
+        
+        # Parse code blocks
+        helpText = helpText.replace('\\\\code', '\n:::{code-block} lua\n')
```
Comment (alexanderbock) [inline generatedocs.py:196]:
```suggestion
        helpText = helpText.replace("\\\\code", "\n:::{code-block} lua\n")
```

File: generatedocs.py [hunk 16]
```diff
@@ -0,0 +1,413 @@
...
+################################################################################
+#                         SCRIPTING API HELPER FUNCTIONS                       #
+################################################################################
+
+# This function modifies the scripting library so that the doxygen 
+# comments are added to the arguments
+# Supported doxygen parameters: \param \return \code
+def parseDoxygenComments(library):
+    for function in library["functions"]:
+        [helpText, p, returnDescription] = function["help"].partition('\\\\return')
```
Comment (alexanderbock) [inline generatedocs.py:193]:
```suggestion
        [helpText, p, returnDescription] = function["help"].partition("\\\\return")
```

File: generatedocs.py [hunk 17]
```diff
@@ -0,0 +1,413 @@
...
+    
+    # Search pass 1: look up folder “assets/<category>/<assetcomponentname>/”
+    # and see if it exists. If it does, add all files in that folder (no subdirectories)
+    assetDirectory = os.path.join(examplesFolder, (os.path.join(category, name)).lower())
+    if os.path.exists(assetDirectory):
+        filenames = assetsInPath(assetDirectory)
+        examples = []
+        for filename in filenames:
+            # We search for the asset component <name> or "<name>"
+            regex = r'\b' + name + r'\b|\b\"' + name + r'\"\b'
```
Comment (alexanderbock) [inline generatedocs.py:145]:
```suggestion
            regex = r"\b" + name + r"\b|\b\"" + name + r"\"\b"
```

File: generatedocs.py [hunk 18]
```diff
@@ -0,0 +1,413 @@
...
+def findShortestAssetInPath(path, name):
+    assetFiles = assetsInPathRecursive(path)
+    # Sort by shortest asset file first                
+    # This will ensure the simplest asset is displayed
+    assetFiles.sort(key=getFileLength)
+
+    # Find example for the asset component
+    for assetFile in assetFiles:   
+        # Search for Type = "<name>"
+        regex = r'Type = \"' + name + r'\"'
```
Comment (alexanderbock) [inline generatedocs.py:123]:
```suggestion
        regex = r"Type = \"" + name + r"\""
```

File: generatedocs.py [hunk 19]
```diff
@@ -0,0 +1,413 @@
...
+# matches occured, as well as the header if specified
+def getLinesAndContentFromFile(assetFile, regex, lookForHeader = False):
+    lines = []
+    isHeaderComment = True
+    header = ""
+    content = ""
+    description = ""
+    luaComment = "-- "
+    headerFinished = 0
+    with open(assetFile, 'r', encoding='utf8') as file:
```
Comment (alexanderbock) [inline generatedocs.py:81]:
```suggestion
    with open(assetFile, "r", encoding="utf8") as file:
```

File: generatedocs.py [hunk 20]
```diff
@@ -0,0 +1,413 @@
...
+    git.checkout(branchName, "--", dataAssetsPath)
+    print("Done cloning assets folder from OpenSpace repository")
+    assetsFolderPath = os.path.abspath(os.path.join(folderName, dataAssetsPath))
+    return assetsFolderPath
+
+# Returns the number of lines in a file
+def getFileLength(path):
+    try:
+        # Read binary file as it is faster and we only want to know the length
+        with open(path, 'rb') as fp:
```
Comment (alexanderbock) [inline generatedocs.py:43]:
```suggestion
        with open(path, "rb") as fp:
```

File: generatedocs.py [hunk 21]
```diff
@@ -0,0 +1,413 @@
...
+
+################################################################################
+#                      ASSET COMPONENTS HELPER FUNCTIONS                       #
+################################################################################
+
+# Clones asset directory from OpenSpace git repository
+# Uses the latest master
+# Returns absolute path to the asset folder
+def cloneAssetsFolderGit(folderName):
+    def printProgress(op_code, cur_count, max_count=None, message=''):
```
Comment (alexanderbock) [inline generatedocs.py:16]:
```suggestion
    def printProgress(op_code, cur_count, max_count=None, message=""):
```

File: generatedocs.py [hunk 22]
```diff
@@ -0,0 +1,413 @@
...
+    # Create index file
+    indexAssetComponentsTemplate = environment.get_template("indexAssetComponentsTemplate.txt")
+    outputIndex = indexAssetComponentsTemplate.render(assetCategories=assetCategories)
+    with open(assetsOutputPath + '/index.md', 'w') as f:
+        f.write(outputIndex)
+
+    # Print out missing assets
+    total = noOfFoundAssets + len(componentsMissingAssets)
+    percentFound = noOfFoundAssets / (total) * 100
+    print('\n\n')
```
Comment (alexanderbock) [inline generatedocs.py:308]:
```suggestion
    print("\n\n")
```

File: generatedocs.py [hunk 23]
```diff
@@ -0,0 +1,413 @@
...
+    print("Number of found asset examples (ignoring base classes):")
+    print(noOfFoundAssets, "of", total, "or", "%.1f" % percentFound, "%\n")
```
Comment (alexanderbock) [inline generatedocs.py:310]:
```suggestion
    print(f"{noOfFoundAssets} of {total} or {"%.1f" % percentFound}%\n"")
```

File: generatedocs.py [hunk 24]
```diff
@@ -0,0 +1,413 @@
...
+    line = "-" * 80 
+    print(line)
+    print(len(componentsMissingAssets), "asset components are missing example files:")
+    print(line) 
+    print('\n'.join(componentsMissingAssets).join("\n\n\n"))
```
Comment (alexanderbock) [inline generatedocs.py:315]:
```suggestion
    print("\n".join(componentsMissingAssets).join("\n\n\n"))
```

File: generatedocs.py [hunk 25]
```diff
@@ -0,0 +1,413 @@
...
+################################################################################
+
+def generateScriptingApi(environment, outputFolder, folderNameScripting, jsonLocation):
+    # Create target folder
+    scriptingOutputPath = os.path.join(outputFolder, folderNameScripting)
+    if not os.path.exists(scriptingOutputPath):
+        os.mkdir(scriptingOutputPath)
+
+    # Opening JSON file
+    f = open(os.path.join(jsonLocation, 'scriptingApi.json'))
```
Comment (alexanderbock) [inline generatedocs.py:329]:
```suggestion
    f = open(os.path.join(jsonLocation, "scriptingApi.json"))
```

File: generatedocs.py [hunk 26]
```diff
@@ -0,0 +1,413 @@
...
+    # Convert JSON String to Python Dictionary
+    scriptingApi = json.load(f)
+
+    # Create the pages for the scripting libraries
+    scriptingApiTemplate = environment.get_template("scriptingApiTemplate.txt")
+    for library in scriptingApi:
+        # Go through all the functions in that library and print out a md file
+        library = parseDoxygenComments(library)
+        outputLibrary = scriptingApiTemplate.render(library=library)
+        with open(os.path.join(scriptingOutputPath, library["fullName"]+'.md'), 'w') as f:
```
Comment (alexanderbock) [inline generatedocs.py:340]:
```suggestion
        with open(os.path.join(scriptingOutputPath, library["fullName"] + ".md"), "w") as f:
```

File: generatedocs.py [hunk 27]
```diff
@@ -0,0 +1,413 @@
...
+            f.write(outputLibrary)
+
+    # Create index file
+    indexScriptingTemplate = environment.get_template("indexScriptingTemplate.txt")
+    outputIndex = indexScriptingTemplate.render(libraries=scriptingApi)
+    with open(os.path.join(scriptingOutputPath, "index.md"), 'w') as f:
```
Comment (alexanderbock) [inline generatedocs.py:346]:
```suggestion
    with open(os.path.join(scriptingOutputPath, "index.md"), "w") as f:
```

File: generatedocs.py [hunk 28]
```diff
@@ -0,0 +1,413 @@
...
+        f.write(outputIndex)
+
+################################################################################
+#                         CREATE RENDERABLE OVERVIEW                           #
+################################################################################
+
+def generateRenderableOverview(environment, outputFolder, folderNameAssets, jsonLocation):
+    # Open JSON file
+    f = open(os.path.join(jsonLocation, 'assetComponents.json'))
```
Comment (alexanderbock) [inline generatedocs.py:355]:
```suggestion
    f = open(os.path.join(jsonLocation, "assetComponents.json"))
```

File: generatedocs.py [hunk 29]
```diff
@@ -0,0 +1,413 @@
...
+                    images[assetComponent["name"]] = image
+                renderables.append(assetComponent)
+
+    # Create overview file
+    renderableOverviewTemplate = environment.get_template("renderableOverviewTemplate.txt")
+    outputOverview = renderableOverviewTemplate.render(folderNameAssets=folderNameAssets,
+                                                       renderables=renderables, 
+                                                       images=images
+                                                       )
+    with open(os.path.join(outputFolder, "renderableOverview.md"), 'w') as f:
```
Comment (alexanderbock) [inline generatedocs.py:381]:
```suggestion
    with open(os.path.join(outputFolder, "renderableOverview.md"), "w") as f:
```

File: generatedocs.py [hunk 30]
```diff
@@ -0,0 +1,413 @@
...
+
+def generateIndexFile(environment, outputFolder, folderNameScripting, folderNameAssets):
+    indexTemplate = environment.get_template("indexTemplate.txt")
+
+    outputIndex = indexTemplate.render(
+        folderNameScripting=folderNameScripting, 
+        folderNameAssets=folderNameAssets,
+        renderableOverview="renderableOverview"    
+    )
+    with open(os.path.join(outputFolder, "index.md"), 'w') as f:
```
Comment (alexanderbock) [inline generatedocs.py:396]:
```suggestion
    with open(os.path.join(outputFolder, "index.md"), "w") as f:
```

File: generatedocs.py [hunk 31]
```diff
@@ -0,0 +1,413 @@
...
+        f.write(outputIndex)
+
+jsonLocation="json"
+outputFolder="generated"
+folderNameAssets="assetComponents"
+folderNameScripting="scriptingApi"
```
Comment (alexanderbock) [inline generatedocs.py:402]:
```suggestion
jsonLocation = "json"
outputFolder = "generated"
folderNameAssets = "assetComponents"
folderNameScripting = "scriptingApi"
```

**Comment by alexanderbock** [comment]
C:\Development\OpenSpace-Docs (releases/v0.20.0)> .\make.bat html
Running Sphinx v7.2.6
No module named 'git'
loading pickled environment... done

File: generatedocs.py [hunk 32]
```diff
@@ -0,0 +1,413 @@
...
+# Find all .asset files in the root dir and subdirs
+def assetsInPathRecursive(root):
+    filenames = []
+
+    for path, subdirs, files in os.walk(root):
+        for file in files:
+            if file.endswith(".asset"):
+                filenames.append(os.path.join(path, file))
+
+    return filenames
```
Comment (GPayne) [inline generatedocs.py:62]: I suggest no blank lines in a short function like this. The python style guide recommends keeping spaces within a function to a minimum, or you could also have two spaces between functions that are outside of a class.

File: generatedocs.py [hunk 33]
```diff
@@ -0,0 +1,413 @@
...
+        if not file.readable():
+            return None
+        # Read file line by line
+        for l_no, line in enumerate(file, 1):
+            # Check for header comment at top of file 
+            if not line.startswith(luaComment):
+                isHeaderComment = False
+            
+            # If we are in header comment, split into header 
+            # and description
```
Comment (GPayne) [inline generatedocs.py:91]: this comment can fit in a single line

File: generatedocs.py [hunk 34]
```diff
@@ -0,0 +1,413 @@
...
+
+    # Convert JSON String to Python Dictionary
+    assetCategories = json.load(f)
+    
+    images = {}
+    renderables = []
+    for category in assetCategories:
+        if category["name"] == "Renderable" or category["name"] == "ScreenSpaceRenderable":
+            for assetComponent in category["classes"]:
+                if assetComponent["name"] == category["name"]:
```
Comment (GPayne) [inline generatedocs.py:365]: Alternate suggestion:

```
                if assetComponent["name"] != category["name"]:
                    # Find example image
                    image = findAssetScreenshot(assetComponent["name"])
                    if image:
                        images[assetComponent["name"]] = image
                    renderables.append(assetComponent)
```

**Review by GPayne — APPROVED** [review]

File: using-openspace/scripting/index.md (hunk 8 above)
Comment (ylvaselling) [reply using-openspace/scripting/index.md:2]: Yes, this PR does not add any hand written documentation

File: using-openspace/navigation/joystick.md [hunk 35]
```diff
@@ -1,5 +1,5 @@
...
...ssible to navigate using a controller, such as an Xbox controller, PS4 controller, or a SpaceMouse. If you want to use any of these supported controllers you need to include their corresponding `.asset` file in a profile. See the manual on [Profiles](../../content/profiles) for further information about this. All of the joystick assets are located in the sub-folder `data/assets/util/joysticks` inside the OpenSpace folder. It is important to include the right asset file for the type of controller you are using. For example, the wireless Xbox asset will not work with a non-wireless Xbox controller and vice versa. If you are unsure about what type of controller you are using, you could instead try to include the `any-joystick` asset. This asset will try to auto-detect what type of controller is connected to the computer and add the corresponding asset file automatically. If you want to use a controller that does not have an already provided asset file in OpenSpace, you can create your own asset file for it, see [Joystick Customization](joystick-customization) for an in-depth guide. The same page also includes information on how to customize the asset files to enable new functionality.
```
[the remark this answers is not in the document]
Comment (ylvaselling) [reply using-openspace/navigation/joystick.md:2]: Just the links, no there isn't

File: generatedocs.py (hunk 34 above)
Comment (ylvaselling) [reply generatedocs.py:365]: I would prefer to not add another layer of nesting

File: generatedocs.py [hunk 36]
```diff
@@ -0,0 +1,413 @@
...
+        return { "header" : header, "description": description, "content" : content, "lines" : lines }
+    else: 
+        return None
+
+# Takes an asset component name and matches it to all files in a 
+# path, and return the shortest asset with matches.
+def findShortestAssetInPath(path, name):
+    assetFiles = assetsInPathRecursive(path)
+    # Sort by shortest asset file first                
+    # This will ensure the simplest asset is displayed
```
[the remark this answers is not in the document]
Comment (ylvaselling) [reply generatedocs.py:118]: No, the comment is not outdated as we use the shortest one when we don't find an exact match in the examples folder

File: generatedocs.py (hunk 4 above)
Comment (ylvaselling) [reply generatedocs.py:353]: Because this file is placed in the "base" folder in the output directory

image

File: generatedocs.py (hunk 12 above)
Comment (ylvaselling) [reply generatedocs.py:1]: Are you running the python virtual environment with `.\.venv\Scripts\activate` ?

**Comment by ylvaselling** [comment]
> 👍 Some comments aside from the one in code:
> 
>     1. Did not manage to enable dark mode, even if I put my browser in dark mode (Chrome).
> 
>     2. Building the page creates a lot of .md files that now appear as changed files. These should be added to .gitignore
> 
>     3. Is this considered the finished-finished version for 0.20.0?  Should it relaly be merged into the `master` branch then?
> 
>     4. Also, regarding the placeholder pages, if this is the finished-finished version then we should try to fill a few of them with content.  Like the `Community` page
> 
>     5. The order of the navigation list changes when I navigate through the page. EI have seen "About" move to the top of the list if I for example navigate to "Getting Started -> Hardware requirements"
> 
>     6. A not for future work: We still need to go through all the pages and make sure we use consistent title casing everywhere. E.g. note how "Handling Errors" and "Hardware requirements" under getting started has different title casing
> 
>     7. Design wise, it would be nice to have some more indentaiton fo rhte subitems in the navigation menu to the left. See exmaple below .It's difficult to see that Development is a category under "Contributing". (This image also shows some structural issues that we should fix...)
>        image
>        Version indentation on Godot page, which to me is clearer and easier to distinguish visually, for comparison
>        image
> 
>     8. On my screen resolution, when viewed in 100%, I find the webpage a bit difficult to read due to the text taking up all the width. How about reducing the width of the page content, like for example Godot does?  This makes the content much easier to overview and read. But we could probably make it a little bit wider than them. https://docs.godotengine.org/en/stable/about/faq.html
>        
>        * See pic below for how it looks on my display. Also note the double "Navigation" in the Getting started menu (these are two different pages, but looks a bit weird IMO)
>        
>        image

1. You can try out the dark mode in chrome with the little paint brush in the inspector 
image
2. 👍 
3. I would assume the master branch should always display the latest version?
4. Agree
5. Can't reproduce this issue. Have you cleared your browser history and run make clean in case you have built the documentation before?
6. 👍 
7. Agree, I think there is some more design work to be done. Adding it to the list
8. For future work I'd like to include a right hand side table of contents view. This would mitigate this issue

**Comment by ylvaselling** [comment]
> For 1, I think having a button somewhere to toggle between light, dark, and auto mode similar to the current version would be best I believe
> 
>     3. We'd want to merge this branch into `master` but then either keep this branch around or recreate it once we push the version out.  Functionally the same as leaving it as a separate branch, but getting some comments on this PR is worth the little detour I think.  So no, it's not the finished-finished version yet, but a big step towards it
> 
>     4. That was my desire to use 100% of the screen width. If that is too wide, then the browser should be thinner instead ;)  I don't really like how "modern" webpage waste a lot of the screen space.

1. I agree with this, but maybe we can add this at a later stage where we also include the right hand side floating panel with the table of contents?
2. 👍 
3. 👍

File: generatedocs.py (hunk 12 above)
Comment (ylvaselling) [reply generatedocs.py:1]: I have updated the Readme with the proper way to run it

File: conf.py [hunk 37]
```diff
@@ -1,5 +1,11 @@
 import os
 
+# opening file_1.py and reading it with read() and executing if with exec()
```
Comment (alexanderbock) [inline conf.py:3]:
```suggestion
# Generate the files that dynamically depend on asset files in the main OpenSpace repo
```

Comment (alexanderbock) [reply conf.py:3]: Not sure about the formulation, but something more specific than currently

File: generatedocs.py (hunk 12 above)
Comment (alexanderbock) [reply generatedocs.py:1]: Works great now!

File: generatedocs.py [hunk 38]
```diff
@@ -0,0 +1,424 @@
...
+  if not origin:
+    print("No remote origin found. Creating OpenSpace remote...")
+    origin = repo.create_remote("origin", "https://github.com/OpenSpace/OpenSpace")
+
+  print(f"Fetching OpenSpace on branch {branch_name}... this might take a while")
+  origin.fetch(progress=print_progress)
+
+  data_assets_path = "data/assets" 
+  git = repo.git()
+  git.checkout(branch_name, "--", data_assets_path)
```
Comment (alexanderbock) [inline generatedocs.py:36]:
```suggestion
  print("Cloning assets examples...")
  repo = Repo.init(folder_name)

  # Create a new remote if there isn't one already created
  origin = repo.remotes[0] if len(repo.remotes) > 0 else None
  if not origin:
    print("No remote origin found. Creating OpenSpace remote...")
    origin = repo.create_remote("origin", "https://github.com/OpenSpace/OpenSpace")

  print(f"Fetching OpenSpace... this might take a while")
  origin.fetch(progress=print_progress, depth=1)

  data_assets_path = "data/assets"
  git = repo.git()
  git.checkout("--", data_assets_path)
```

The `fetch` will default to the `master` branch, so no need to specify that separately.  The big change is the addition of `depth=1` which will make a shallow copy that speeds up the fetch dramatically

File: generatedocs.py [hunk 39]
```diff
@@ -0,0 +1,424 @@
...
+    # Read binary file as it is faster and we only want to know the length
+    with open(path, "rb") as fp:
+      if fp.readable():
+        return len(fp.read()) 
+      else:
+        return None
+  except IOError:
+    print(f"Could not open file path {path}")
+    return None
+
```
Comment (alexanderbock) [inline generatedocs.py]:
```suggestion
```

Function replaced with a lambda below

File: generatedocs.py [hunk 40]
```diff
@@ -0,0 +1,424 @@
...
+def find_shortest_asset_in_path(path, name):
+  """
+  Takes an asset component name and matches it to all files in a 
+  path, and return the shortest asset with matches.
+  """
+  asset_files = assets_in_path_recursive(path)
+  # Sort by shortest asset file first                
+  # This will ensure the simplest asset is displayed
+  asset_files.sort(key=get_file_length)
```
Comment (alexanderbock) [inline generatedocs.py:128]:
```suggestion
  asset_files.sort(key=lambda file: os.stat(file).st_size)
```

**Review by alexanderbock — APPROVED** [review]
